episodes/excthreesteps.Rmd

```diff
@@ -0,0 +1,613 @@
+    
+    
+    # if elastic like regularisation used smooth the displacement field
+    if sigma_elastic > 0:
+        disp_field_x = gaussian_filter(disp_field_x, sigma_elastic, mode='nearest')
+        disp_field_y = gaussian_filter(disp_field_y, sigma_elastic, mode='nearest')
+    
+    # update deformation field from disp field
+    def_field[:, :, 0] = disp_field_x + X
+    def_field[:, :, 1] = disp_field_y + Y
+            
+    # transform the image using the updated deformation field
+    warped_image = resampImageWithDefField(source, def_field)
+
+    # update images if required for this iteration
+    if disp_freq > 0 and it % disp_freq == 0:
+        plt.figure(3)
+        dispImage(warped_image)
+        plt.pause(0.05)
+        plt.figure(4)
+        plt.clf()
+        dispDefField(def_field, spacing=disp_spacing, plot_type=disp_method_df)
+        plt.xlim(x_lims)
+        plt.ylim(y_lims)
+        plt.pause(0.05)
+        plt.figure(5)
+        plt.clf()
+        up_field_to_display = scale_update_for_display * np.dstack((update_x, update_y))
+        up_field_to_display += np.dstack((X, Y))
+        dispDefField(up_field_to_display, spacing=disp_spacing, plot_type=disp_method_up)
+        plt.xlim(x_lims)
+        plt.ylim(y_lims)
+        plt.pause(0.05)
+    
+    # calculate MSD between target and warped image
+    MSD = calcMSD(target, warped_image)
+
+    # display numerical results
+    print('Level {0:d}, Iteration {1:d}: MSD = {2:f}\n'.format(lev, it, MSD))
+    
+    # check for improvement in MSD if required
+    if check_MSD and MSD >= prev_MSD:
+        # restore previous results and finish level
+        def_field = def_field_prev
+        warped_image = resampImageWithDefField(source, def_field)
+        print('No improvement in MSD')
+        break
+    
+    # update previous values of def_field and MSD
+    def_field_prev = def_field.copy()
+    prev_MSD = MSD.copy()
+```
+
+## 6. Final Results Display
+
+After completing the iterative registration process, the final step is to display the warped image and the deformation field. This provides a visual representation of the registration outcome.
+
+```python
+# Final Results: Display warped image and deformation field
+plt.figure(3)
+dispImage(warped_image)
+plt.figure(4)
+plt.clf()
+dispDefField(def_field, spacing=disp_spacing, plot_type=disp_method_df)
+plt.xlim(x_lims)
+plt.ylim(y_lims)
+plt.figure(5)
+plt.clf()
+up_field_to_display = scale_update_for_display * np.dstack((update_x, update_y))
+up_field_to_display += np.dstack((X, Y))
+dispDefField(up_field_to_display, spacing=disp_spacing, plot_type=disp_method_up)
+plt.xlim(x_lims)
+plt.ylim(y_lims)
+```
+
+
+## 7. Return Results
+
+The final step of the registration function is to return the warped image and the deformation field. These outputs can be further analysed or used for downstream processing as needed.
+
+```python
+return warped_image, def_field
+```
+
+
+
+# Full code

```

Comment:
Once adopting Python PROJECT_SLUG conventions, can we please point users to this new path `src/mirc/utils/demonsReg.py` for full code
```
PROJECT_SLUG
├── ...
├── README.md
├── pyproject.toml
├── src
│   └── PROJECT_SLUG
│       └── PROJECT_SLUG.py
└── tests
    └── test_dummy.py
```